### PR TITLE
Proposal: provide a regex safe string constant containing all emojis

### PR DESCRIPTION
**Pull request by black-puppydog**
This is something I made to scratch my own itch: a `str` constant that holds a `regex`-safe concatenation of all emojis. This wasn't *completely* trivial to make. I spent way too much time figuring out this here:

```
002A FE0F 20E3 *️⃣  ← Emoji
002A 20E3      *⃣   ← Emoji
002A           *   ← bog-standard ASCII asterisk
```

To clarify what the issue here is: the code point for the ascii asterisk (`*`) is contained in the two emojis above. Of course, passing these to a regex naively can fail or (worse) silently corrupt the regex.

I currently implemented this as a feature in this crate and copied the resulting string into my own code, but I figured this might be of use to someone else

I'm just unsure where it should live. I could polish this a bit (mostly squashing and updating the docs and such) if you think this crate here might be a good place, but I could also just make a new crate just for this...

### CODE REVIEW

**Comment by black-puppydog** [comment]
ah, there's one unrelated commit in here that renames all `.tpl` files to `.tera`. I didn't really find much on this; the tera website's examples just use the `.html` extension when generating html. But the tera extension for vscode recognizes `.tera` while ignoring `.tpl` so I made this commit as a convenince to myself. This commit can obviously disappear without much discussion.

File: emojic-gen/templates/regex.tera [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//! This module contains the same set of emojis as the [`crate::grouped`] module, but
+//! without the sub modules. This make it a bit more messy, but allows for shorter
+//! references from code.
+//!
+//! # Examples
+//!
+//! ```rust
+//! // prints: 🖼️
+//! println!("{}", emojic::flat::FRAMED_PICTURE);
+//! ```
```
Comment (black-puppydog) [inline emojic-gen/templates/regex.tera:20]: This is the documentation I'd have to update of course. 😆

**Comment by Cryptjar** [comment]
Hi, you bring up a very good point. Trying to do regex parsing with emojis seems to have quite a few pitfalls. I think there should be at least a paragraph of caution about using emoji-based regexes, e.g. in the `emojic::text` module doc.

Just out of curiosity, what is your use-case?
Do you just want to check whether a string contains any emoji, or do you want to remove them or something?

If I think about use-cases, I mostly imagine things like replacing just specific emojis or maybe something like a reverse `emojic::text::parse_text`. In both cases, I imagine something more specific like an additional `regex_escaped_grapheme` field on the `Emoji` struct would be more useful.

BTW: If you coincidentally want to just detect emojis, why not doing something "simple" like:

```rust
regex::Regex::new(
    &emojic::grouped::base_emojis()
        .map(|e| regex::escape(&e.grapheme))
        .collect::<Vec<_>>()
        .join("|")
)
```

On the other hand, if you want to replace things, it seems you forgot about ordering, because if I try to adapte your test case towards replacement:

```rust
let re = regex::Regex::new(&format!("({})", EMOJI_REGEX))?;
let output = re.replace("\u{1F469}\u{1F3FB}\u{200D}\u{2764}\u{200D}\u{1F48B}\u{200D}\u{1F468}\u{1F3FF}", "#");
```

I just get the following `output`:

```text
👩#\u{200d}❤\u{200d}💋\u{200d}👨🏿
\u{1F469}#\u{200D}\u{2764}\u{200D}\u{1F48B}\u{200D}\u{1F468}\u{1F3FF}
```

Apparently, the Fitzpatrick modifier took precedence over the longer emoji sequence.

But even with `replace_all` not all emojis are replaced, which is very strange:

```text
👩#\u{200d}❤\u{200d}#\u{200d}👨#
\u{1F469}#\u{200D}\u{2764}\u{200D}#\u{200D}\u{1F468}#
```

Emoji regexes are really complicated.

**Comment by Cryptjar** [comment]
I digged a bit deeper, first, adding all emoji variants (using `grouped::all_variants()`) and then adding ordering (it seems like a good start to just sort by UTF-8 encoding size), it improved, but it did not work completely, yet. Then I double-checked the emoji, which we already have in the crate, I thought (I was wrong):

> ```rust
> println!("{}", KISS.gender(Pair::Mixed).tone((Tone::Light, Tone::Dark))); // 👩🏻‍❤️‍💋‍👨🏿 (`U+1F469 U+1F3FB U+200D U+2764 U+FE0F U+200D U+1F48B U+200D U+1F468 U+1F3FF`)
> ```
> _from the docs <https://docs.rs/emojic/latest/emojic/grouped/people_and_body/family/static.KISS.html>_

Notice that the comment mentions an additional `U+FE0F` (the [emoji variant selector](https://en.wikipedia.org/wiki/Emoji#Emoji_versus_text_presentation)), which makes e.g. a textual ❤︎ into an emoji ❤️. You apparently did not get that emoji from `emojic`.  Any way, when using all emojis, ordering and matching against emojis with emoji variant selector, it works 🎉

```rust
// All "proper" emojis
let mut all_graphemes = Vec::from_iter(
    grouped::all_variants()
        .flatten()
        .map(|e| ::regex::escape(&e.grapheme)),
);
// Sort them by size, works here, it might fall apart in some edge cases tho
all_graphemes.sort_by_key(|g| -(g.len() as i128));
// Use as regex
let re = ::regex::Regex::new(&format!("({})", all_graphemes.join("|")))?;
// Match agains a "proper" emoji (notice the variant selector "fe0f")
let output = re.replace_all("\u{1F469}\u{1F3FB}\u{200D}\u{2764}\u{fe0f}\u{200D}\u{1F48B}\u{200D}\u{1F468}\u{1F3FF}", "#");
```

It finally yields the expected `output`:

```text
#
```

However, this nicely highlight further difficulties with matching emojis, because I think it is quite reasonable to output emojis with the variant selector (to reduce rendering ambiguities), but when matching emojis, one wants to (potentially) accept both (with and without variant selector).

---
**Update:** actually, one can very well argue that the above regex (modulo maybe the sorting strategy) is Unicode conform, since it only considers those emojis with emoji variant selectors (called _fully-qualified_ emoji). It should also match all RGI sequence, since the RGI list is our source list. The Unicode standard explicitly allows "implementations" to choose to ignore emoji ZWJ sequence that do not contain (sufficient) emoji variant selectors (called _minimally-qualified_ and _unqualified_) when processing emojis \[[ref](https://unicode.org/reports/tr51/#Emoji_Variation_Selector_Notes)\].

**Comment by black-puppydog** [comment]
Thanks for the long and detailed reply! 

> Just out of curiosity, what is your use-case?

I am trying to re-create the JS hashtag parser from [svbergerem/markdown-it-hashtag](https://github.com/svbergerem/markdown-it-hashtag) which in turn uses [hashtag-regex](https://github.com/mathiasbynens/hashtag-regex/). `hashtag-regex` takes its [emoji definitions](https://github.com/mathiasbynens/hashtag-regex/blob/fb654e6d864b41050db3d18dc3723ea9b64dfbe7/script/inject-sequences.js#L7-L11) from the `unicode` package.

Since I'm not an expert at either JS nor Rust I looked around for a crate that does something similar and stumbled upon emojic. 🙂 
So my application is really detection only (and extraction, but that's the same really).
Specifically, I want to use a single `lazy_static` regex to detect all hashtags, just like in the `hashtag-regex` crate. This is what I currently use:

```rust
&format!("^[#\u{FE5F}\u{FF03}][\\*\\+\\\x2D0-9A-Z_a-z({})]+\\b", EMOJI_REGEX)
```

Where `EMOJI_REGEX` is the concatenation of all emoji variants, joined with `|`.

---------------------------

> You apparently did not get that emoji from emojic

I'm assuming you mean the `FE0F` or the `002A` one? In any case, the way I came up with this triade of asterisks is that my first attempt at building my regex was to just concatenate all emojis, without the call to `regex::escape` (which I didn't even know about then) and the call to `Regex::new` would fail with an error about a `*` without preceding characters. Which was actually lucky; if the `*` code point had not been at the beginning of an emoji, directly after the `|`, it might have just silently built the wrong regex! 😛 
After a bit of searching I found out which emojis caused the problem and built the example I gave by copying the code points from the emoji definitions on unicode.org.
So, short answer: yes I did (kinda) get these code points from `emojic`.

---------------------------

The code snippets you provided work for me!
I'd still argue that they're non-trivial and boiler-plate-y. This is what I ended up with in my code:

```rust
use lazy_static::lazy_static;
use regex::Regex;

fn generate_emoji_regex() -> String {
    // All "proper" emojis
    let mut all_graphemes = Vec::from_iter(
        emojic::grouped::all_variants()
            .flatten()
            .map(|e| ::regex::escape(&e.grapheme)),
    );
    // Sort them by size, works here, it might fall apart in some edge cases tho
    all_graphemes.sort_by_key(|g| -(g.len() as i128));
    format!("({})", all_graphemes.join("|"))
}

lazy_static! {
    static ref ALL_EMOJI_VARIANTS: String = generate_emoji_regex();
    static ref HASHTAG_RE_PREFIX: Regex = Regex::new(&format!("^[#\u{FE5F}\u{FF03}][\\*\\+\\\x2D0-9A-Z_a-z({})]+\\b", ALL_EMOJI_VARIANTS.to_string())).unwrap();
    static ref HASHTAG_RE_EXACT:  Regex = Regex::new(&format!("^[#\u{FE5F}\u{FF03}][\\*\\+\\\x2D0-9A-Z_a-z({})]+\\b", ALL_EMOJI_VARIANTS.to_string())).unwrap();
}
```

If I could just `use emojic::regex::ALL_EMOJI_VARIANTS_RE;` or such that would be perfect. 😁

**Comment by black-puppydog** [comment]
Okay, this finally resulted in a _very_ simple solution. Only took me a week and a half. 😆 
In the end I'm using emojic only for the tests, but I couldn't have done this without your insight. 

This has been a pretty nice learning experience, thanks! 😊

**Pull request closed without merging**